ecs/pod_creator.go

```diff
@@ -27,11 +31,235 @@ func NewBasicECSPodCreator(c cocoa.ECSClient, v cocoa.Vault) (*BasicECSPodCreato
 	return nil, errors.New("TODO: implement")
 }
+
+// exportTags converts strings into ECS tags.
+func exportTags(tags []string) []*ecs.Tag {
+	ecsTags := []*ecs.Tag{}
+
+	for _, tag := range tags {
+		ecsTag := &ecs.Tag{}
+		ecsTag.SetKey(tag)
+		ecsTags = append(ecsTags, ecsTag)
+	}
+
+	return ecsTags
+}
+
+// exportStrategy converts the strategy and parameter into an ECS placement strategy.
+func exportStrategy(strategy *cocoa.ECSPlacementStrategy, param *cocoa.ECSStrategyParameter) []*ecs.PlacementStrategy {
+	placementStrat := ecs.PlacementStrategy{}
+
+	placementStrat.SetField(utility.FromStringPtr(param))
+
+	if strategy == nil {
+		placementStrat.SetType("")
+	} else {
+		placementStrat.SetType(string(*strategy))
+	}
+
+	return []*ecs.PlacementStrategy{&placementStrat}
+}
+
+// exportEnvVars converts the non-secret environment variables into ECS environment variables.
+func (m *BasicECSPodCreator) exportEnvVars(variables []cocoa.EnvironmentVariable) []*ecs.KeyValuePair {
+	keyValuePairs := []*ecs.KeyValuePair{}
+
+	for _, variable := range variables {
+		if variable.SecretOpts == nil {
+			keyValue := ecs.KeyValuePair{}
+			keyValue.SetName(utility.FromStringPtr(variable.Name)).SetValue(utility.FromStringPtr(variable.Value))
+			keyValuePairs = append(keyValuePairs, &keyValue)
+		}
+	}
+
+	return keyValuePairs
+}
+
+// getSecrets retrieves the secrets from the secret environment variables for the pod.
+func (m *BasicECSPodCreator) getSecrets(merged cocoa.ECSPodCreationOptions) []cocoa.SecretOptions {
+	secrets := []cocoa.SecretOptions{}
+
+	for _, def := range merged.ContainerDefinitions {
+		for _, variable := range def.EnvVars {
+			if variable.SecretOpts != nil {
+				secrets = append(secrets, *variable.SecretOpts)
+			}
+		}
+	}
+
+	return secrets
+}
+
+// createSecrets creates secrets that do not already exist.
+func (m *BasicECSPodCreator) createSecrets(ctx context.Context, secrets []cocoa.SecretOptions) error {
+	if m.vault == nil {
+		return errors.New("vault does not exist")
+	}
+
+	for _, secret := range secrets {
+		if !*secret.Exists {
+			_, err := m.vault.CreateSecret(ctx, secret.PodSecret.NamedSecret)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// translatePodSecrets translates secret options into pod secrets.
+func translatePodSecrets(secrets []cocoa.SecretOptions) []cocoa.PodSecret {
+	podSecrets := []cocoa.PodSecret{}
+
+	for _, secret := range secrets {
+		podSecrets = append(podSecrets, secret.PodSecret)
+	}
+
+	return podSecrets
+}
+
+// exportSecrets converts secret options into ECS Secrets.
+func exportSecrets(secrets []cocoa.SecretOptions) []*ecs.Secret {
+	ecsSecrets := []*ecs.Secret{}
+
+	for _, secret := range secrets {
+		ecsSecret := ecs.Secret{}
+		ecsSecret.SetName(utility.FromStringPtr(secret.NamedSecret.Name))
```

Comment:
The [ECS docs on secret](https://docs.aws.amazon.com/sdk-for-go/api/service/ecs/#Secret) suggest that this is not quite what you want. The meaning of "name" in the ECS secret is that this is the environment variable name you'll be exposing (which is `(cocoa.EnvironmentVariable).Name`). In contrast, `NamedSecret.Name` is the friendly alias for the secret in Secrets Manager. Furthermore, the [docs on Secret](https://docs.aws.amazon.com/sdk-for-go/api/service/ecs/#Secret) also state that the `ValueFrom` has to be a reference to the full ARN of the secret in Secrets Manager (i.e. the result from `Vault.CreateSecret`), so `ValueFrom` basically acts as a pointer to tell ECS where to retrieve the secret.

If you're creating a new secret for the caller, you should preserve the result from `Vault.CreateSecret` then set it as the secret's `Name`. If using an existing secret, you should assume the user will give you a valid ARN as the secret name. That way, you can do `SetValueFrom(utility.FromStringPtr(secret.Name)` to use the secret name as an ARN.

You can see an example [here (under Example Container Definitions)](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/specifying-sensitive-data-secrets.html).

As an aside, you can just use `secret.Name`. Since `PodSecret` is embedded in the `cocoa.SecretOptions` and `NamedSecret` is embedded in `PodSecret`, you can use `(cocoa.SecretOptions).Name`.

Reply:
Ah okay, that makes sense. I was getting a bit confused about the multiple meanings of the same word, but this clears it up. Thanks!